### PR TITLE
azurerm_cognitive_account supports identity storage disable_local_auth fqdns public_network_access_enabled and restrict_outbound_network_access

### PR DESCRIPTION
**Pull request by ms-henglu**
`azurerm_cognitive_account` supports `identity` `storage` `disable_local_auth` `fqdns` `public_network_access_enabled` and `restrict_outbound_network_access`

The tests are listed as the followings.

```

=== RUN   TestAccCognitiveAccount_basic
=== PAUSE TestAccCognitiveAccount_basic
=== CONT  TestAccCognitiveAccount_basic
--- PASS: TestAccCognitiveAccount_basic (178.22s)
=== RUN   TestAccCognitiveAccount_speechServices
=== PAUSE TestAccCognitiveAccount_speechServices
=== CONT  TestAccCognitiveAccount_speechServices
--- PASS: TestAccCognitiveAccount_speechServices (164.60s)
=== RUN   TestAccCognitiveAccount_speechServicesWithStorage
=== PAUSE TestAccCognitiveAccount_speechServicesWithStorage
=== CONT  TestAccCognitiveAccount_speechServicesWithStorage
--- PASS: TestAccCognitiveAccount_speechServicesWithStorage (207.76s)
=== RUN   TestAccCognitiveAccount_requiresImport
=== PAUSE TestAccCognitiveAccount_requiresImport
=== CONT  TestAccCognitiveAccount_requiresImport
--- PASS: TestAccCognitiveAccount_requiresImport (184.34s)
=== RUN   TestAccCognitiveAccount_complete
=== PAUSE TestAccCognitiveAccount_complete
=== CONT  TestAccCognitiveAccount_complete
--- PASS: TestAccCognitiveAccount_complete (181.70s)
=== RUN   TestAccCognitiveAccount_update
=== PAUSE TestAccCognitiveAccount_update
=== CONT  TestAccCognitiveAccount_update
--- PASS: TestAccCognitiveAccount_update (257.20s)
=== RUN   TestAccCognitiveAccount_qnaRuntimeEndpoint
=== PAUSE TestAccCognitiveAccount_qnaRuntimeEndpoint
=== CONT  TestAccCognitiveAccount_qnaRuntimeEndpoint
--- PASS: TestAccCognitiveAccount_qnaRuntimeEndpoint (244.40s)
=== RUN   TestAccCognitiveAccount_qnaRuntimeEndpointUnspecified
=== PAUSE TestAccCognitiveAccount_qnaRuntimeEndpointUnspecified
=== CONT  TestAccCognitiveAccount_qnaRuntimeEndpointUnspecified
--- PASS: TestAccCognitiveAccount_qnaRuntimeEndpointUnspecified (91.48s)
=== RUN   TestAccCognitiveAccount_cognitiveServices
=== PAUSE TestAccCognitiveAccount_cognitiveServices
=== CONT  TestAccCognitiveAccount_cognitiveServices
--- PASS: TestAccCognitiveAccount_cognitiveServices (154.87s)
=== RUN   TestAccCognitiveAccount_withMultipleCognitiveAccounts
=== PAUSE TestAccCognitiveAccount_withMultipleCognitiveAccounts
=== CONT  TestAccCognitiveAccount_withMultipleCognitiveAccounts
=== CONT  TestAccCognitiveAccount_withMultipleCognitiveAccounts
--- PASS: TestAccCognitiveAccount_withMultipleCognitiveAccounts (109.40s)
=== RUN   TestAccCognitiveAccount_networkAcls
=== PAUSE TestAccCognitiveAccount_networkAcls
=== CONT  TestAccCognitiveAccount_networkAcls
--- PASS: TestAccCognitiveAccount_networkAcls (345.53s)
=== RUN   TestAccCognitiveAccount_identity
=== PAUSE TestAccCognitiveAccount_identity
=== CONT  TestAccCognitiveAccount_identity
--- PASS: TestAccCognitiveAccount_identity (472.47s)

=== RUN   TestAccCognitiveAccount_metricsAdvisor
=== PAUSE TestAccCognitiveAccount_metricsAdvisor
=== CONT  TestAccCognitiveAccount_metricsAdvisor
--- PASS: TestAccCognitiveAccount_metricsAdvisor (502.44s)

```

### CODE REVIEW

File: azurerm/internal/services/cognitive/cognitive_account_resource.go [hunk 1]
```diff
@@ -104,6 +108,83 @@ func resourceCognitiveAccount() *pluginsdk.Resource {
...
+			},
+
+			"custom_subdomain_name": {
+				Type:         pluginsdk.TypeString,
+				Optional:     true,
+				ForceNew:     true,
+				ValidateFunc: validation.StringIsNotEmpty,
+			},
+
+			"disable_local_auth": {
```
Comment (katbyte) [inline azurerm/internal/services/cognitive/cognitive_account_resource.go:132]: could we change this to be more consistant with the provider
```suggestion
			"local_auth_enabled": {
```

File: azurerm/internal/services/cognitive/cognitive_account_resource.go [hunk 2]
```diff
@@ -104,6 +108,83 @@ func resourceCognitiveAccount() *pluginsdk.Resource {
 				}, false),
 			},
 
+			"aad_client_id": {
+				Type:         pluginsdk.TypeString,
+				Optional:     true,
+				ForceNew:     true,
+				ValidateFunc: validation.StringIsNotEmpty,
```
Comment (katbyte) [inline azurerm/internal/services/cognitive/cognitive_account_resource.go:115]: could we properly validate this?

File: azurerm/internal/services/cognitive/cognitive_account_resource.go [hunk 3]
```diff
@@ -104,6 +108,83 @@ func resourceCognitiveAccount() *pluginsdk.Resource {
...
+			},
+
+			"aad_tenant_id": {
+				Type:         pluginsdk.TypeString,
+				Optional:     true,
+				ForceNew:     true,
+				ValidateFunc: validation.StringIsNotEmpty,
```
Comment (katbyte) [inline azurerm/internal/services/cognitive/cognitive_account_resource.go:122]: could we properly validate this?

File: azurerm/internal/services/cognitive/cognitive_account_resource.go [hunk 4]
```diff
@@ -104,6 +108,83 @@ func resourceCognitiveAccount() *pluginsdk.Resource {
...
+				Optional: true,
+				Default:  false,
+			},
+
+			"fqdns": {
+				Type:     pluginsdk.TypeList,
+				Optional: true,
+				Elem: &pluginsdk.Schema{
+					Type:         pluginsdk.TypeString,
+					ValidateFunc: validation.StringIsNotEmpty,
```
Comment (katbyte) [inline azurerm/internal/services/cognitive/cognitive_account_resource.go:143]: we cann properly check this is a FQDN?

File: azurerm/internal/services/cognitive/cognitive_account_resource.go [hunk 5]
```diff
@@ -146,7 +227,46 @@ func resourceCognitiveAccount() *pluginsdk.Resource {
...
 			},
 
-			"custom_subdomain_name": {
+			"public_network_access_enabled": {
+				Type:     pluginsdk.TypeBool,
+				Optional: true,
+				Default:  true,
+			},
+
+			"restrict_outbound_network_access": {
```
Comment (katbyte) [inline azurerm/internal/services/cognitive/cognitive_account_resource.go:236]: could we put restrict on the end to be consistant with other bools?
```suggestion
			"outbound_network_access_restrited": {
```

File: azurerm/internal/services/cognitive/cognitive_account_resource.go [hunk 6]
```diff
@@ -146,7 +227,46 @@ func resourceCognitiveAccount() *pluginsdk.Resource {
...
+				Optional: true,
+				Default:  false,
+			},
+
+			"storage": {
+				Type:     pluginsdk.TypeList,
+				Optional: true,
+				Elem: &pluginsdk.Resource{
+					Schema: map[string]*pluginsdk.Schema{
+						"resource_id": {
```
Comment (katbyte) [inline azurerm/internal/services/cognitive/cognitive_account_resource.go:247]: what type of resource is this?

File: azurerm/internal/services/cognitive/cognitive_account_resource.go [hunk 7]
```diff
@@ -146,7 +227,46 @@ func resourceCognitiveAccount() *pluginsdk.Resource {
...
+						"identity_client_id": {
+							Type:         pluginsdk.TypeString,
+							Optional:     true,
+							ValidateFunc: validation.IsUUID,
+						},
+					},
+				},
+			},
+
+			"super_user": {
```
Comment (katbyte) [inline azurerm/internal/services/cognitive/cognitive_account_resource.go:262]: is this user name?
```suggestion
			"super_user_name": {
```

File: website/docs/r/cognitive_account.html.markdown [hunk 8]
```diff
@@ -56,6 +60,22 @@ The following arguments are supported:
...
+  
+* `identity` - (Optional) An `identity` block is documented below.
+
+* `public_network_access_enabled` - (Optional) Whether public network access is allowed for the Cognitive Account. Defaults to `true`.
+
+* `restrict_outbound_network_access` - (Optional) Whether outbound network access is restricted for the Cognitive Account. Defaults to `false`.
+
+* `storage` - (Optional) An `identity` block is documented below.
+
+* `super_user` - (Optional) The super user of Metrics Advisor. This attribute is only set when kind is `MetricsAdvisor`. Changing this forces a new resource to be created.
```
Comment (katbyte) [inline website/docs/r/cognitive_account.html.markdown:75]: should this be
```suggestion
* `metrics_advisor_super_user_name` - (Optional) The super user of Metrics Advisor. This attribute is only set when kind is `MetricsAdvisor`. Changing this forces a new resource to be created.
```

File: website/docs/r/cognitive_account.html.markdown [hunk 9]
```diff
@@ -44,10 +44,14 @@ The following arguments are supported:
...
+* `kind` - (Required) Specifies the type of Cognitive Service Account that should be created. Possible values are `Academic`, `AnomalyDetector`, `Bing.Autosuggest`, `Bing.Autosuggest.v7`, `Bing.CustomSearch`, `Bing.Search`, `Bing.Search.v7`, `Bing.Speech`, `Bing.SpellCheck`, `Bing.SpellCheck.v7`, `CognitiveServices`, `ComputerVision`, `ContentModerator`, `CustomSpeech`, `CustomVision.Prediction`, `CustomVision.Training`, `Emotion`, `Face`,`FormRecognizer`, `ImmersiveReader`, `LUIS`, `LUIS.Authoring`, `MetricsAdvisor`, `Personalizer`, `QnAMaker`, `Recommendations`, `SpeakerRecognition`, `Speech`, `SpeechServices`, `SpeechTranslation`, `TextAnalytics`, `TextTranslation` and `WebLM`. Changing this forces a new resource to be created.
 
 * `sku_name` - (Required) Specifies the SKU Name for this Cognitive Service Account. Possible values are `F0`, `F1`, `S`, `S0`, `S1`, `S2`, `S3`, `S4`, `S5`, `S6`, `P0`, `P1`, and `P2`.
 
+* `aad_client_id` - (Optional) The Azure AD Client ID (Application ID). This attribute is only set when kind is `MetricsAdvisor`. Changing this forces a new resource to be created.
```
Comment (katbyte) [inline website/docs/r/cognitive_account.html.markdown:51]: could we validate that when kind is MetricsAdvisor this is set? also given this is just for mstrics advisor:
```suggestion
* `metrics_advisor_aad_client_id` - (Optional) The Azure AD Client ID (Application ID). This attribute is only set when kind is `MetricsAdvisor`. Changing this forces a new resource to be created.
```

**Review by katbyte — CHANGES_REQUESTED** [review]
Thanks @ms-henglu - i've left some comments inline. we also have some failing tests:
image

File: azurerm/internal/services/cognitive/cognitive_account_resource.go (hunk 6 above)
Comment (ms-henglu) [reply azurerm/internal/services/cognitive/cognitive_account_resource.go:247]: storage account id, I think it's better to rename it to `storage_account_id`. I'll change that.

File: website/docs/r/cognitive_account.html.markdown (hunk 9 above)
Comment (ms-henglu) [reply website/docs/r/cognitive_account.html.markdown:51]: It's not required for metrics advisor, so I think it's not necessary to validate that when kind is MetricsAdvisor this is set.

File: azurerm/internal/services/cognitive/cognitive_account_resource.go (hunk 4 above)
Comment (ms-henglu) [reply azurerm/internal/services/cognitive/cognitive_account_resource.go:143]: fqdn is quite complicated to validate, maybe we can skip it?

**Comment by ms-henglu** [comment]
@katbyte , Hi , I've added a commit to rename these fields. And about the failed test, the cause is that QnAMaker and SpeechServices are not available in `West Europe`, so I run the tests again in `West US`.
image

File: azurerm/internal/services/cognitive/cognitive_account_resource.go (hunk 4 above)
Comment (katbyte) [reply azurerm/internal/services/cognitive/cognitive_account_resource.go:123]: sure

**Comment by katbyte** [comment]
@ms-henglu - that is not going to fix it for the nightly runs. is it availible in any of these regions?
image

**Comment by ms-henglu** [comment]
@katbyte , for `SpeechServices`, it's available in `eastus2` and `francecentral`, so I fixed it. But `QnAMaker` is only available in `west us`, so I hardcoded it in tests and added comments for it.

Only one test failed in tc, because it needs to register a subscription through[ this form](https://forms.office.com/Pages/ResponsePage.aspx?id=v4j5cvGGr0GRqy180BHbR3k4-f_9d9xPhWkgOUub9YhUNDZJUERRVVVDME4xNDVNRjEwMTNZV1dHSS4u). Could you please help me with it? Because I can't see the subscriptionId.

tc tests.
image

local tests with same paramters.
image

**Review by katbyte — APPROVED** [review]
Thanks @ms-henglu - all tests pass and this LGTM 🙌

**Pull request merged**